rfcs/1478-variables.md

```diff
@@ -0,0 +1,123 @@
+
+### Variable names
+
+A variable is a named value. In the Seq query language, variable names start with `$`, so that there's no risk of getting incorrect results when the name of a variable conflicts with the name of an event property or column (neither of these things can use names that start with `$`).
+
+Formally, as a regular expression, variables names match the pattern `\$[A-Za-z_0-9]+`.
+
+Examples of variable names:
+
+ * `$a`
+ * `$TEST`
+ * `$_`
+ * `$1`
+ * `$event99`
+
+### Variable values
+
+Variables can hold any Seq runtime value. For practical purposes, this means JSON-style `null`, Booleans, numbers, strings, arrays, and objects can all be stored in a variable.
+
+### Variable scope
+
+Variables are defined and used within a single Seq browser window. Variables are not persisted or tracked by the Seq server.
+
+The _Variables_ tool window (see below) provides a simple "copy" function that can be used to replicate a variable from one Seq browser session into another.
+
+### Setting, editing, and clearing variables
+
+#### In the _Variables_ tool window
+
+In the right-hand (signal bar) panel, a new tool window called _Variables_ shows the list of variables, and their values, held in the current browser session.
+
+> Some reorganization of the signal bar is planned so that tool windows can be supported.
+
+The _Variables_ tool window has a "+" button, allowing a new variable to be added, alongside features to edit, remove, and clear current variables.
+
+A user of the Seq interface can click the add button, give the variable a name (`$` prefix is enforced), and choose to give the variable a literal text or numeric value.
+
+Variables in the tool window will provide a context menu with _Copy name_, _Copy value_, and _Copy assignment_, where the assignment is a query language statement that sets the variable value.
+
+#### Direcly, in the query language
+
+Variables can be set in the Seq search box using `select into`:
+
+```sql
+select ToIsoString(now()) into $rightNow
+```
+
+The `into` clause can only appear in scalar queries (initially, only those without a `from stream` clause), and only when there is exactly one column in the `select` clause column list.
+
+#### With shortcuts or context menus
+
+Various UI elements will assign values to variables. For example, the green tick drop-down beside property values will include an _Assign to variable_ item that will assign the property's value to a variable.
+
+When one of these UI elements is clicked, the _Variables_ tool window will show (if hidden), and the new variable will appear with an automatically-generated name, derived from the context that provided the value. The user can choose to click and rename the variable, or accept the system-generated name.
+
+UI elements that may include variable assignments:
+
+ * The "event actions" _Type_ menu
+ * The green tick context menu on properties, including all levels of JSON value viewer
+ * Cells in result set tables (possibly)
```

Comment:
since we can't do:

`select count(*) into $count`

it would be nice to have a contextual way to assign the results of aggregate queries to variables.